packages/account/src/BiconomySmartAccountV2.ts

```diff
@@ -154,33 +238,277 @@ export class BiconomySmartAccountV2 extends BaseSmartAccount {
+          return userOp
+        }
+
+        if (!userOp.callData) {
+          throw new Error('Userop callData cannot be undefined')
+        }
+
+        const account = await this._getAccountContract()
+
+        const decodedDataSmartWallet = account.interface.parseTransaction({
+          data: userOp.callData.toString()
+        })
+        if (!decodedDataSmartWallet) {
+          throw new Error('Could not parse call data of smart wallet for userOp')
+        }
+
+        const smartWalletExecFunctionName = decodedDataSmartWallet.name
+
+        if (smartWalletExecFunctionName === 'executeCall') {
+          Logger.log('originally an executeCall for Biconomy Account')
+          const methodArgsSmartWalletExecuteCall = decodedDataSmartWallet.args
+          const toOriginal = methodArgsSmartWalletExecuteCall[0]
+          const valueOriginal = methodArgsSmartWalletExecuteCall[1]
+          const dataOriginal = methodArgsSmartWalletExecuteCall[2]
+
+          batchTo.push(toOriginal)
+          batchValue.push(valueOriginal)
+          batchData.push(dataOriginal)
+        } else if (smartWalletExecFunctionName === 'executeBatchCall') {
+          Logger.log('originally an executeBatchCall for Biconomy Account')
+          const methodArgsSmartWalletExecuteCall = decodedDataSmartWallet.args
+          batchTo = methodArgsSmartWalletExecuteCall[0]
+          batchValue = methodArgsSmartWalletExecuteCall[1]
+          batchData = methodArgsSmartWalletExecuteCall[2]
+        }
+
+        if (approvalRequest.to && approvalRequest.data && approvalRequest.value) {
+          batchTo = [approvalRequest.to, ...batchTo]
+          batchValue = [approvalRequest.value, ...batchValue]
+          batchData = [approvalRequest.data, ...batchData]
+
+          newCallData = await this.encodeExecuteBatch(batchTo, batchValue, batchData)
+        }
+        let finalUserOp: Partial<UserOperation> = {
+          ...userOp,
+          callData: newCallData
+        }
+
+        // Requesting to update gas limits again (especially callGasLimit needs to be re-calculated)
+        try {
+          finalUserOp = await this.estimateUserOpGas(finalUserOp)
+          const cgl = ethers.BigNumber.from(finalUserOp.callGasLimit)
+          if (finalUserOp.callGasLimit && cgl.lt(ethers.BigNumber.from('21000'))) {
+            return {
+              ...userOp,
+              callData: newCallData
+            }
+          }
+          Logger.log('userOp after estimation ', finalUserOp)
+        } catch (error) {
+          Logger.error('Failed to estimate gas for userOp with updated callData ', error)
+          Logger.log(
+            'sending updated userOp. calculateGasLimit flag should be sent to the paymaster to be able to update callGasLimit'
+          )
+        }
+        return finalUserOp
+      }
+    } catch (error) {
+      Logger.log('Failed to update userOp. sending back original op')
```

Comment:
Correct normal grammer. Make `s` capital in 

**S**ending back original UserOp

Reply:
done